oil_lang/expr_eval.py

```diff
@@ -459,6 +490,351 @@ def _EvalConst(self, node):
     # yet.
     raise AssertionError(id_)
 
+  def _EvalCommandSub(self, node):
+    # type: (command_sub) -> value_t
+
+    id_ = node.left_token.id
+    # &(echo block literal)
+    if id_ == Id.Left_CaretParen:
+      return value.Str('TODO: value.Block')
+    else:
+      stdout = self.shell_ex.RunCommandSub(node)
+      if id_ == Id.Left_AtParen:  # @(seq 3)
+        strs = self.splitter.SplitForWordEval(stdout)
+        return value.MaybeStrArray(strs)
+      else:
+        return value.Str(stdout)
+
+  def _EvalShArrayLiteral(self, node):
+    # type: (sh_array_literal) -> value_t
+    words = braces.BraceExpandWords(node.words)
+    strs = self.word_ev.EvalWordSequence(words)
+    #log('ARRAY LITERAL EVALUATED TO -> %s', strs)
+    return value.MaybeStrArray(strs)
+
+  def _EvalDoubleQuoted(self, node):
+    # type: (double_quoted) -> value_t
+
+    # In an ideal world, I would *statically* disallow:
+    # - "$@" and "${array[@]}"
+    # - backticks like `echo hi`
+    # - $(( 1+2 )) and $[] -- although useful for refactoring
+    #   - not sure: ${x%%} -- could disallow this
+    #     - these enters the ArgDQ state: "${a:-foo bar}" ?
+    # But that would complicate the parser/evaluator.  So just rely on
+    # strict_array to disallow the bad parts.
+    return value.Str(self.word_ev.EvalDoubleQuotedToString(node))
+
+  def _EvalSingleQuoted(self, node):
+    # type: (single_quoted) -> value_t
+    return value.Str(word_compile.EvalSingleQuoted(node))
+
+  def _EvalBracedVarSub(self, node):
+    # type: (braced_var_sub) -> value_t
+    return value.Str(self.word_ev.EvalBracedVarSubToString(node))
+
+  def _EvalSimpleVarSub(self, node):
+    # type: (simple_var_sub) -> value_t
+    return value.Str(self.word_ev.EvalSimpleVarSubToString(node))
+
+  def _EvalUnary(self, node):
+    # type: (expr__Unary) -> value_t
+    child = _PyObjToValue(self._EvalExpr(node.child)) # XXX
+    if node.op.id == Id.Arith_Minus:
+      UP_child = child
+      with tagswitch(child) as case:
+        if case(value_e.Bool):
+          child = cast(value__Bool, UP_child)
+          return value.Int(-child.b) # gets promoted...
+
+        elif case(value_e.Int):
+          child = cast(value__Int, UP_child)
+          return value.Int(-child.i)
+
+        elif case(value_e.Float):
+          child = cast(value__Float, UP_child)
+          return value.Float(-child.f)
```

Comment:
can we put `else: raise AssertionError()` here and in all the `with tagswitch()`?

Or it could be something like `error.Expr(...)`, or maybe `error.InvalidType` which derives from `error.Expr()`